pydantic_ai_slim/pydantic_ai/_cli/web.py

```diff
@@ -0,0 +1,116 @@
+# (includes deprecated tools plus tools needing config like mcp_server and memory)
+UNSUPPORTED_CLI_TOOLS = DEPRECATED_BUILTIN_TOOL_TYPES | frozenset({'mcp_server', 'memory'})
+
+
+# TODO add MCP support issue https://github.com/pydantic/pydantic-ai/issues/3635
+def run_web_command(
+    agent_path: str | None = None,
+    host: str = '127.0.0.1',
+    port: int = 7932,
+    models: list[str] | None = None,
+    tools: list[str] | None = None,
+    instructions: str | None = None,
+) -> int:
+    """Run the web command to serve an agent via web UI.
+
+    If an agent is provided, its model and builtin tools are used as defaults.
+    CLI-specified models and tools are added on top. Duplicates are removed.
+
+    Args:
+        agent_path: Agent path in 'module:variable' format. If None, creates generic agent.
+        host: Host to bind the server to.
+        port: Port to bind the server to.
+        models: List of model strings (e.g., ['openai:gpt-5', 'anthropic:claude-sonnet-4-5']).
+        tools: List of builtin tool IDs (e.g., ['web_search', 'code_execution']).
+        instructions: System instructions. For generic agent (no agent_path), these become
+            the agent's instructions. With an agent, these are passed as extra instructions
+            to each run.
+    """
+    console = Console()
+
+    if agent_path:
+        agent = load_agent(agent_path)
+        if agent is None:
+            console.print(f'[red]Error: Could not load agent from {agent_path}[/red]')
+            return 1
+    else:
+        agent = Agent(instructions=instructions)
+
+    if agent.model is None and not models:
+        console.print('[red]Error: At least one model (-m) is required when agent has no model[/red]')
+        return 1
+
+    # build models list
+    if agent.model is not None:
+        resolved = infer_model(agent.model)
+        agent_model = f'{resolved.system}:{resolved.model_name}'
+        models = [agent_model] + [m for m in (models or []) if m != agent_model]
+
+    # collect builtin tools
+    all_tool_instances: list[AbstractBuiltinTool] = [
+        t
+        for t in agent._builtin_tools  # pyright: ignore[reportPrivateUsage]
+        if isinstance(t, AbstractBuiltinTool)
+    ]
+
+    # then CLI tools
+    if tools:
+        for tool_id in tools:
+            if tool_id in UNSUPPORTED_CLI_TOOLS:
+                console.print(
+                    f'[yellow]Warning: "{tool_id}" requires configuration and cannot be enabled via CLI, skipping[/yellow]'
+                )
+                continue
+            tool_cls = BUILTIN_TOOL_TYPES.get(tool_id)
+            if tool_cls is None:
+                console.print(f'[yellow]Warning: Unknown tool "{tool_id}", skipping[/yellow]')
+                continue
+            all_tool_instances.append(tool_cls())
+
+    # For generic agent, instructions are already in the agent.
+    # For provided agent, pass instructions as extra instructions for each run.
```

Comment:
We can simplify this by never passing them to agent, and always passing them to the agent run, right?

Reply:
you're correct on that